frontend/src/metabase/dashboard/components/grid/utils.js

```diff
@@ -0,0 +1,69 @@
+import _ from "underscore";
+import { Responsive as ResponsiveGrid } from "react-grid-layout";
+import { color } from "metabase/lib/colors";
+
+export function adaptLayoutForBreakpoint({
+  layout,
+  breakpoints,
+  targetBreakpoint,
+  closestBreakpoint,
+  columns,
+  compactType,
+}) {
+  return ResponsiveGrid.utils.findOrGenerateResponsiveLayout(
+    layout,
+    breakpoints,
+    targetBreakpoint,
+    closestBreakpoint,
+    columns,
+    compactType,
+  );
+}
+
+function sumVerticalSpace(layout) {
+  return layout.reduce((sum, current) => sum + current.h, 0);
+}
+
+export function generateMobileLayout({
+  desktopLayout,
+  defaultCardHeight,
+  heightByDisplayType = {},
+}) {
+  const mobile = [];
+  desktopLayout.forEach(item => {
+    const card = item.dashcard.card;
+    const height = heightByDisplayType[card.display] || defaultCardHeight;
+    mobile.push({
+      ...item,
+      x: 0,
+      y: sumVerticalSpace(mobile),
+      h: height,
+      w: 1,
+    });
+  });
+  return mobile;
```

Comment:
Total nit-picking: could this not be a reduce?

Reply:
Do you mean `sumVerticalSpace`? Asking because you selected the `forEach` loop

Reply:
The overall shape of the thing looked like a reduce.

I didn't write it down, but I saw something _roughly_ like this:

```js
  return  desktopLayout.reduce( item => {
    const card = item.dashcard.card; 
    const height = heightByDisplayType[card.display] || defaultCardHeight;    
    
    return {     
      ...item,      
      x: 0,      
      y: sumVerticalSpace(mobile),      
      h: height,      
      w: 1,    
    };
  } , []} 
  ```

Reply:
Not sure I understand how `reduce` fits here (sorry, I'm going to sleep soon, can be lagging a bit)

The main thing here is the `sumVerticalSpace(mobile)` line. This whole function could be a simple `map`, but for each item, in the mobile layout we need to calculate how much vertical space is already taken by the previous items, so they are aligned one after the other vertically. So each loop iteration has to have access to previously created mobile layout items

Reply:
Yeah, no worries! It's pretty good as it is.

Reply:
Since `reduce` gives access to the value you're building via the first arg in the callback, you can technically write the function like this:

```
return desktopLayout.reduce((mobile, item) => {
  const card = item.dashcard.card;
  const height = heightByDisplayType[card.display] || defaultCardHeight;
  return [
    ...mobile,
    {
       ...item,
       x: 0,
       y: sumVerticalSpace(mobile),
       h: height,
       w: 1,
    }
  ];
}, []);
```

Your version is already a well contained pure function so I wouldn't worry too much about it